controllers/remote/cluster.go

```diff
@@ -27,9 +27,11 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
+// ClusterClientGetter returns a new remote client.
+type ClusterClientGetter func(c client.Client, cluster *clusterv1.Cluster, scheme *runtime.Scheme) (client.Client, error)
```

Comment:
How strongly do you feel about this vs just using the inline `func()` definition in the structs where we need it?

Reply:
(I'm not crazy about having this type here but my concern isn't blocking)

Reply:
What are your concerns?

Reply:
Mainly that nobody outside of test could would likely need to use `ClusterClientGetter`

Reply:
So the concern is cluttering up CAPI libraries?

Reply:
I'd prefer the client here as main entrypoint, it looks cleaner when you read the structs

Reply:
Sort of. I'm ok doing things to make testing easier (like having the `remoteClient` field in the reconciler so we can swap in a test-specific implementation). I view this as similar to the view on interfaces (don't force them on your consumers).

Reply:
problem is that if we change the arguments for example, all the function definitions in the structs have to change as well, and we're maintaining these types everywhere instead of one place

Reply:
I think that's a valid concern. My one additional insight here is that this is an exported function and not an interface we're forcing on anyone. If they want to use it they can, if they don't want to use it they don't have to. We aren't exporting an overly large interface which is generally when forcing an interface on clients becomes extremely onerous. I'm fine with either this PR or not, it's not a huge deal either way.